include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,591 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime_api.h>
```

Comment:
If the file includes a CUDA header, possible to make the file name be `memory_pool.cuh`, or will have compile error like : 
```
/HierarchicalKV/include/merlin/memory_pool.hpp:19:10: fatal error: cuda_runtime_api.h: No such file or directory
   19 | #include <cuda_runtime_api.h>
      |          ^~~~~~~~~~~~~~~~~~~~

```

Reply:
But it doesn't contain device code. So you don't need NVCC to compile it. That problem is avoided already when compiling the test solely with GCC by expanding the include search path in the `CMakeLists.txt`. This is the proper way to make it valid C++ compile job.

The main reason for the behavior described by you is that NVCC silently expands the default header search path. The fact that if you `compile a normal C/C++ header with NVCC and GCC leads to different include search path order` avoids many `I cannot compile` questions on StackOverflow, but software-design-wise, it is a somewhat unlucky decision. One could argue that with this decision they traded a typical beginner error for an issue that can cause peculiar results in larger software solutions (compile module that **includes same header** with **different compilers** causes **different dependent header resolution...** ;-) ). The same is true when you trigger linking. A regular C++ compiler/linker like GCC should only mark modules as dependent that are specified by the user. However, NVCC automatically adds CUDA modules to the link dependencies. You have no control about this. With this, you get functions resolved to CUDA functions that you may not want to have resolved. It fixes a beginner issue. But think about situations where you have multiple CUDA frameworks. Suddenly, you get stuff resolved silently in the background instead of getting nice verbose error message.

Hence, one could call it a somewhat poor design decision. As said in the first paragraph, for compiling the test, it is properly resolved by extending the search path. In other words, by adding `$ENV{CUDA_HOME}/include` to the include search path, and explicitly link `libcudart` with the executable.

Reply:
Whatever you plan to compile the lib, the end-users should compile successfully anytime by following the guide in the Readme. Is it right?

Reply:
`But it doesn't contain device code. ` --Yes, so why include the `cuda_runtime_api.h`?

Reply:
It invokes the CUDA runtime API. It doesn't contain `device` code.

Anyway, the NVCC issue was discussed. I fixed the file ending according to your wishes. Please check and resolve conversation if happy now.